Lightning+Lightrec updates and other optimizations

ZachCook opened the pull request:
~~Split out cpu_lightrec.cpp from cpu.cpp instead of being mashed together.~~

Added hugepage, codebuffer, mmap at 0x0, and SP GP hit RAM optimizations and fallbacks, see commits for more details.

Use lightrec's GTE registers directly when not using beetle interpreter (while maintaining backwards compatibility with savestates).

Removed the Cycle Timing Check dynarec option.

Fixes #942

ZachCook committed:
Lightrec: Move and update lightning and lightrec include files

Moved all include files from lightning-lightrec-include/ to include/

Moved lightning.h to lightning-actual.h and made lightning.h include
lightning-actual.h and define HAVE_MMAP for lightning, instead of defining
in the Makefile which applied it to other parts of beetle, causing issues
on Windows in libretro-common/vfs

ZachCook committed:
lightrec: Adjust to lightrec API changes and HAVE_MMAP for lightning

Fix bios freeze with lightrec, AssertIRQ must change lightrec's CAUSE too

cleanup unused cop0 functions

Fix build with updated lightrec

Add lightrec missing build file

Lightrec: Remove kunseg for pgxp rw, since lightrec_rw does that now

Lightrec: Fix PGXP crash by adding LWU/SWU meta-opcodes

Otherwise lightrec uses the default when PGXP needs the nonhw functions

Lightrec: allow threaded recompiler in debug builds

ZachCook committed:
Lightrec: enable codebuffer, memfd, hugepage, and rewrite mmap

Codebuffer is much faster due to lightrec's 32-bit LUT optimization

lightrec: attempt memfd and hugepages for ram mmap

lightrec_init_mmap: Clean and rewrite io_base mmap code

Fix Windows and cleanup error handling in init_mmap

Replaces supported_io_bases array with min/inc/max_io_base which makes it easier to avoid ranges of address for different platforms

Moves actual mapping to lightrec_try_map, so it can be called a second time with different address ranges

Fix unmap of lightrec codebuffer

Add THP based madvise hugepage fallback

If /sys/kernel/mm/transparent_hugepage/shmem_enabled is set to advise
then memfd shared mapping for ram and mirrors can use a 2MB hugepage

Only try either hugepage mmap or not, and only show warning if number
of maps was incorrect

Clean up some header includes

ZachCook committed:
Change GTE to allow using lightrec's registers

ZachCook committed:
Remove "Cycle Timing Check" (execute_one) dynarec option, use new lightrec api

This was only useful for debugging, users don't benefit from it, the config
will fallback to using the beetle interpreter if it was selected

The lightrec execute_one function this option called was removed, for
equivalent debugging pass current timestamp instead of next_event_ts to
lightrec_execute

Use lightrec cop2_notify for PGXP, update interpreter to new api

add cpu_timestamp hack to so lightrec tracks more cycles

add out of memory exit condition

clean and optimize execution loop

Copy all CP0 regs

ZachCook committed:
support MainRAM at 0x0, use INVALID_PTR instead of NULL

This allows less code to be emitted by lightrec, but will only
be used if root user enables it with sysctl -w vm.mmap_min_addr=0

pcercuei commented on Makefile.common:
```diff
@@ -375,7 +376,8 @@ ifeq ($(HAVE_LIGHTREC), 1)
                   $(DEPS_DIR)/lightrec/reaper.c \
-                  $(DEPS_DIR)/lightrec/regcache.c
+                  $(DEPS_DIR)/lightrec/regcache.c \
+                  $(DEPS_DIR)/lightrec/tlsf/tlsf.c
```
You shouldn't have to compile TLSF since you disabled the code buffer support (ENABLE_CODE_BUFFER=0)

ZachCook replied:
ENABLE_CODE_BUFFER=1 was set in 3012b97d6e2ec3bcd6cec4f8ed7e9bf22f568ef8 of this PR, there were many changes squashed together, and it could maybe use even more to make this more clear maybe?

pcercuei replied:
Yes I saw it after my message. I guess this change should be squashed into 3012b97 then.

Ryunam commented:
After testing this with a few games on my Windows 11 system, I have noticed a few Runahead-related issues:

- As reported on Discord, enabling Runahead in RetroArch with either single instance or secondary instance mode and having the dynarec active seems to be triggering unpredictable crashes with this PR, at several spots. This is with the same values/settings used with current master. Also, the amount of dynarec cycles do not appear to affect the likelihood of the crashes.
- Based on my testings so far, the crashes can occur either at startup, especially if one fastforwards right after launching the content, or when loading a save. The latter in particular is being triggered here consistently whenever I try to load my "Akumajou Dracula X: Gekka no Yasoukyoku" save from the memory card screen with Runahead active, which worked normally before this PR. Again, fastforwarding seems to greatly increase the chance of RetroArch and the core suddenly crashing with this setup.
- If Dynarec SPU Samples is set to any value above 1 and Runahead is enabled with Single Instance mode, the audio will get noticeably choppy and distorted. Disabling Runahead or setting it to Second Instance Mode makes it sound normal again.

LibretroAdmin commented:
@Ryunam Hi, are your issues addressed now?

Ryunam commented:
Okay, I have performed a few more tests after the last commit:

- The audio issue caused by the combination of SPU Samples set to a value higher than 1 and Runhead Single Instance mode seems to be fixed. I tried switching the values back and forth during gameplay and everything appears to sound normally.
- However, the Dynarec SPU Samples by itself (with or without Runahead) seems not to play nicely with a few games. "Final Fantasy" (the PS1 remake of FF1) stood out during my testing, as any value higher than 1 triggers missing and sustained notes in the music of that game. A clear example can be seen by starting the game, then immediately entering the first city, Corneria. Set the samples to 16 and the music will start having missing parts or oddly prolonged notes. Setting the Samples value back to 1 makes the music play normally again. I assume this is partly expected behavior, as the feature in question might not work well with certain games depending on how the music files are streamed, but i figure it's worth pointing it out.
- Interestingly, the crashes do not seem to happen anymore either! I have tried to trigger them multiple times and so far so good.

ZachCook commented:
@Ryunam 
I'm surprised the crashes disappeared, though I couldn't reproduce so just let me know if they come back.

The SPU Samples option was expected to cause that, I'll add a note in the description about it, it's basically just running the sampling at a lower frequency and then running multiple times to catch up, it's a performance hack, I'm surprised it works as well as it does. Though now that you found a game I can test I might be able to improve it further by doing some run time detection or tweaking the timing code, but that may be a bigger project for the future.

Ryunam commented:
Sorry @ZachCook, I stand corrected. The crashes appeared once again while I was testing other games, so they seem to be unrelated to the SPU changes. 
They are also very unpredictable: sometimes the affected games will launch normally and then crash when loading a save, sometimes the crash will occur almost immediately at startup, some other times they won't happen at all.

One consistent situation where I can reproduce the crash right now is with the Japanese PS1 port of Street Fighter Zero 2:

- Under Core Options -> Emulation Hacks, set Skip BIOS to ON.
- Launch Street Fighter Zero 2 and hold the Fastforward button right after starting the game. One thing that may also contribute to the crash being triggered is pressing/mashing the Circle button while it's fastforwarding, but this was not verified.
- On my system the game and RA will crash either after the Capcom logo animation, or later on when selecting an option from the main menu (for example, "Versus Mode").
- All of the above is with Runahead enabled as a per-core override (activated at startup automatically) with Second Instance ON.

The core as compiled from current master before this PR never triggers any crash with these same steps and/or settings.

LibretroAdmin commented:
Hi there, what is the state of this PR as it stands? Are there any roadblocks?

ZachCook commented:
> Hi there, what is the state of this PR as it stands? Are there any roadblocks?

I'll update it again to latest lightrec+lightning, and see if I can reproduce the above mentioned crash, or at least narrow down what triggers it, I wouldn't really want to merge this until that is dealt with.

ZachCook committed:
git subrepo pull deps/lightrec

subrepo:
  subdir:   "deps/lightrec"
  merged:   "dfa18558"
upstream:
  origin:   "https://github.com/pcercuei/lightrec.git"
  branch:   "master"
  commit:   "dfa18558"
git-subrepo:
  version:  "0.4.9"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "ea10886"

ZachCook commented:
Updated to recently released Lightning 2.2.0 and Lightrec 0.6

Still working on crashes with runahead and loading, the hack I had there (restart entire dynarec and switch to interpreter for a while) looks awfully suspect and I'm working on a proper fix to at least rule that out as a cause if it doesn't just fix it.

LibretroAdmin commented:
@ZachCook Is this ready for merging now? Will this have any known regressions vs. the current state of things?

ZachCook commented:
> @ZachCook Is this ready for merging now? Will this have any known regressions vs. the current state of things?

@LibretroAdmin It is not ready yet, I am seeing lots of crashing now than can be triggered by any kind of run-ahead or loading a save state.

You could apply just 8814bcbc9c8fdc552287097e87312ded80bd412a for now, it is quite a useful improvement by itself.

LibretroAdmin commented:
OK, I cherry-picked that commit for now and pushed it to master

ZachCook committed:
Add Dynarec SP GP hit RAM optimization option

Disabled by default for max compatibility
Enabling should generate faster code for most normal games that
do not have sp and gp registers pointing outside of RAM

ZachCook commented:
Still have to re-fix the savestate issue right after bios since I ripped out the previous next_interpreter hack, but otherwise I have tried to update and clean up the code a bit.

I think some of the older commits are a bit messy/WIP that are fixed later in the series since there was such a long period of time since I last was working on this, so it might be easier to view the changes as a whole for some of it.

ZachCook committed:
git subrepo commit (merge) deps/lightning

subrepo:
  subdir:   "deps/lightning"
  merged:   "f4d3464e"
upstream:
  origin:   "https://git.savannah.gnu.org/git/lightning.git"
  branch:   "master"
  commit:   "a6bb2b5a"
git-subrepo:
  version:  "0.4.9"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "ea10886"

ZachCook committed:
Add Parrel Port Mirror support for PS2/PS3 BIOS

LibretroAdmin commented:
The best I have been able to do is update lightning and pull in one commit from here.

cpu.cpp and gte.cpp got refactored completely into cpu.c and gte.c. 
So @ZachCook if you are up for it, I would highly appreciate a new PR that reintegrates the new Lightrec into the current codebase. And this time we actually merge it.

I have tried seeing if I can backport it over but it's too rife with potential pitfalls for someone that did not write the code himself
